cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/common/StructuredRecordStringConversion.java

```diff
@@ -0,0 +1,439 @@
+      return Double.valueOf(key);
+    }
+    throw new IOException("Unable to convert string to type " + type);
+  }
+
+  private static StructuredRecord readRecord(JsonReader reader, Schema schema) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+
+    reader.beginObject();
+    while (reader.peek() != JsonToken.END_OBJECT) {
+      Schema.Field field = schema.getField(reader.nextName());
+      if (field == null) {
+        // Ignore unrecognized fields
+        reader.skipValue();
+        continue;
+      }
+
+      builder.set(field.getName(), readJson(reader, field.getSchema()));
+    }
+    reader.endObject();
+
+    return builder.build();
+  }
+
+  private static Object readUnion(JsonReader reader, Schema unionSchema) throws IOException {
+    JsonToken token = reader.peek();
+    // Based on the token to guess the schema
+    for (Schema schema : unionSchema.getUnionSchemas()) {
+      if (SCHEMA_TO_JSON_TYPE.get(schema.getType()) == token) {
+        return readJson(reader, schema);
+      }
+    }
+
+    throw new IOException("No matching schema found for union type: " + unionSchema);
+  }
+
+
+  private static void writeJson(JsonWriter writer, Schema schema, Object value) throws IOException {
+    switch (schema.getType()) {
+      case NULL:
+        writer.nullValue();
+        break;
+      case BOOLEAN:
+        writer.value((Boolean) value);
+        break;
+      case INT:
+      case LONG:
+      case FLOAT:
+      case DOUBLE:
+        writer.value((Number) value);
+        break;
+      case BYTES:
+        writeBytes(writer, value);
+        break;
+      case STRING:
+        writer.value((String) value);
+        break;
+      case ENUM:
+        writer.value(((Enum) value).name());
+        break;
+      case ARRAY:
+        writeArray(writer, schema.getComponentSchema(), value);
+        break;
+      case MAP:
+        writeMap(writer, schema.getMapSchema(), value);
+        break;
+      case RECORD:
+        writeRecord(writer, schema, value);
+        break;
+      case UNION:
+        writeJson(writer, findUnionSchema(schema, value), value);
+        break;
+    }
+  }
+
+  private static void writeBytes(JsonWriter writer, Object value) throws IOException {
+    if (value instanceof ByteBuffer) {
+      writeBytes(writer, (ByteBuffer) value);
+    } else if (value.getClass().isArray() && value.getClass().getComponentType().equals(byte.class)) {
+      byte[] bytes = (byte[]) value;
+      writeBytes(writer, bytes, 0, bytes.length);
+    } else {
+      throw new IOException("Expects either ByteBuffer or byte[]. Got " + value.getClass());
+    }
+  }
+
+  private static void writeBytes(JsonWriter writer, ByteBuffer buffer) throws IOException {
+    if (buffer.hasArray()) {
+      writeBytes(writer, buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
+    } else {
+      byte[] buf = new byte[buffer.remaining()];
```

Comment:
Should be able to simplify this by just using `Bytes#getBytes(ByteBuffer)`, right?